src/test/script_p2sh_tests.cpp

```diff
@@ -368,8 +390,63 @@ BOOST_AUTO_TEST_CASE(AreInputsStandard)
     txToNonStd2.vin[0].scriptSig << std::vector<unsigned char>(twentySigops.begin(), twentySigops.end());
 
-    BOOST_CHECK(!::AreInputsStandard(CTransaction(txToNonStd2), coins));
+    BOOST_CHECK_EQUAL(Solver(txFrom.vout[6].scriptPubKey, vSolutions), TxoutType::SCRIPTHASH);
+    BOOST_CHECK(!::AreInputsStandard(CTransaction(txToNonStd2), coins, reason, debug));
+    BOOST_CHECK_EQUAL(reason, "bad-txns-input-scriptcheck-sigops");
+    BOOST_CHECK_EQUAL(debug, "input 0: 20 > 15");
     BOOST_CHECK_EQUAL(GetP2SHSigOpCount(CTransaction(txToNonStd2), coins), 20U);
+
+    // TxoutType::NONSTANDARD
+    CMutableTransaction txToNonStd3;
+    txToNonStd3.vout.resize(1);
+    txToNonStd3.vout[0].scriptPubKey = GetScriptForDestination(PKHash(key[1].GetPubKey()));
+    txToNonStd3.vout[0].nValue = 1000;
+    txToNonStd3.vin.resize(1);
+    txToNonStd3.vin[0].prevout.n = 7;
+    txToNonStd3.vin[0].prevout.hash = txFrom.GetHash();
+    txToNonStd3.vin[0].scriptSig << std::vector<unsigned char>(no_sigops.begin(), no_sigops.end());
+
+    BOOST_CHECK_EQUAL(Solver(txFrom.vout[7].scriptPubKey, vSolutions), TxoutType::NONSTANDARD);
+    BOOST_CHECK(!::AreInputsStandard(CTransaction(txToNonStd3), coins, reason, debug));
+    BOOST_CHECK_EQUAL(reason, "bad-txns-input-script-nonstandard");
+    BOOST_CHECK_EQUAL(debug, "input 0");
+
+    static const unsigned char op_return[] = { OP_RETURN }; // SCRIPT_ERR_OP_RETURN
+
+    CMutableTransaction txToNonStd4;
+    txToNonStd4.vout.resize(1);
+    txToNonStd4.vout[0].scriptPubKey = GetScriptForDestination(PKHash(key[1].GetPubKey()));
+    txToNonStd4.vout[0].nValue = 1000;
+    txToNonStd4.vin.resize(1);
+    txToNonStd4.vin[0].prevout.n = 8;
+    txToNonStd4.vin[0].prevout.hash = txFrom.GetHash();
+    txToNonStd4.vin[0].scriptSig = CScript(op_return, op_return + sizeof(op_return));
+
+    // out args for EvalScript
+    std::vector<std::vector<unsigned char>> stack;
+    ScriptError serror;
+
+    BOOST_CHECK_EQUAL(Solver(txFrom.vout[8].scriptPubKey, vSolutions), TxoutType::SCRIPTHASH);
+    BOOST_CHECK(!EvalScript(stack, txToNonStd4.vin[0].scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker(), SigVersion::BASE, &serror));
+    BOOST_CHECK_EQUAL(serror, SCRIPT_ERR_OP_RETURN);
+    BOOST_CHECK(!::AreInputsStandard(CTransaction(txToNonStd4), coins, reason, debug));
+    BOOST_CHECK_EQUAL(reason, "bad-txns-input-scriptsig-failure");
+    BOOST_CHECK_EQUAL(debug, "input 0: OP_RETURN was encountered");
+
+    // TxoutType::WITNESS_UNKNOWN
+    CMutableTransaction txWitnessUnknown;
+    txWitnessUnknown.vout.resize(1);
+    txWitnessUnknown.vout[0].scriptPubKey = GetScriptForDestination(PKHash(key[1].GetPubKey()));
+    txWitnessUnknown.vout[0].nValue = 1000;
+    txWitnessUnknown.vin.resize(1);
+    txWitnessUnknown.vin[0].prevout.n = 9;
+    txWitnessUnknown.vin[0].prevout.hash = txFrom.GetHash();
+    txWitnessUnknown.vin[0].scriptSig << std::vector<unsigned char>(witnessUnknown.begin(), witnessUnknown.end());
```

Comment:
In ba538fba899482f5549a6377d657d9d89b13bfd5 "Propagate validation debug information from AreInputsStandard"

nit: setting `scriptSig` is unnecessary as the thing being checked is the scriptPubKey of the output being spent.